File: Components/SlabParser.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text.RegularExpressions;
using Grasshopper.Kernel;
using Grasshopper.Kernel.Data;
using Grasshopper.Kernel.Types;
using Rhino.Geometry;
using Newtonsoft.Json.Linq;

namespace Enzyme.Components
{
    public class SlabParserComponent : GH_Component
    {
        public SlabParserComponent()
          : base("Slab JSON Parser", "Slab_Parser",
              "Parses Slab JSON data.",
              "Enzyme", "Masterplan")
        {
        }

        public override GH_Exposure Exposure => GH_Exposure.primary;

        protected override void RegisterInputParams(GH_InputParamManager pManager)
        {
            pManager.AddTextParameter("Slab_JSON", "Slab_JSON", "JSON string for slabs", GH_ParamAccess.item, "");
            pManager.AddTextParameter("Filter_Building", "Filter_Building", "Building filters", GH_ParamAccess.list);
            pManager[1].Optional = true;
            pManager.AddTextParameter("Filter_Tower", "Filter_Tower", "Tower filters", GH_ParamAccess.list);
            pManager[2].Optional = true;
            pManager.AddIntegerParameter("Filter_Level", "Filter_Level", "Level filters", GH_ParamAccess.list);
            pManager[3].Optional = true;
            pManager.AddBooleanParameter("ExactMatch", "ExactMatch", "Use exact matching", GH_ParamAccess.item, false);
        }

        protected override void RegisterOutputParams(GH_OutputParamManager pManager)
        {
            pManager.AddCurveParameter("SlabBounds", "SlabBounds", "Slab boundary curves", GH_ParamAccess.tree);
            pManager.AddNumberParameter("Areas", "Areas", "Slab areas", GH_ParamAccess.tree);
            pManager.AddTextParameter("Labels", "Labels", "Slab labels", GH_ParamAccess.tree);
        }

        protected override void SolveInstance(IGH_DataAccess DA)
        {
            Stopwatch sw = Stopwatch.StartNew();
            
            string jsonIn = "";
            DA.GetData(0, ref jsonIn);

            List<string> rawBuilding = new List<string>();
            DA.GetDataList(1, rawBuilding);
            List<string> fBldg = rawBuilding.Where(b => !string.IsNullOrEmpty(b)).ToList();

            List<string> rawTower = new List<string>();
            DA.GetDataList(2, rawTower);
            List<string> fTower = rawTower.Where(t => !string.IsNullOrEmpty(t)).ToList();

            List<int> fLevel = new List<int>();
            DA.GetDataList(3, fLevel);

            bool exactToggle = false;
            DA.GetData(4, ref exactToggle);

            GH_Structure<GH_Curve> outBounds = new GH_Structure<GH_Curve>();
            GH_Structure<GH_Number> outAreas = new GH_Structure<GH_Number>();
            GH_Structure<GH_String> outLabels = new GH_Structure<GH_String>();

            if (string.IsNullOrWhiteSpace(jsonIn))
            {
                this.Message = this.NickName + "\nTime: 0 ms\n---\nAwaiting Data";
                DA.SetDataTree(0, outBounds);
                DA.SetDataTree(1, outAreas);
                DA.SetDataTree(2, outLabels);
                return;
            }

            try
            {
                JObject data = JObject.Parse(jsonIn);
                int bldgIndex = 0;
                int matchCount = 0;

                foreach (var bldg in data)
                {
                    string bldgName = bldg.Key;
                    if (!IsMatch(bldgName, fBldg, exactToggle)) continue;

                    JArray slabs = bldg.Value as JArray;
                    if (slabs == null) continue;

                    int slabIndex = 0;
                    foreach (JToken slabToken in slabs)
                    {
                        JObject slab = slabToken as JObject;
                        if (slab == null) continue;

                        int lvl = slab["floor_index"]?.Value<int>() ?? -1;
                        string towerId = slab["tower_id"]?.Value<string>() ?? "Unknown";

                        if (fLevel.Count > 0 && !fLevel.Contains(lvl)) 
                        {
                            slabIndex++;
                            continue;
                        }
                        if (!IsMatch(towerId, fTower, exactToggle))
                        {
                            slabIndex++;
                            continue;
                        }

                        GH_Path path = new GH_Path(bldgIndex, slabIndex);
                        string label = $"{bldgName} | {towerId} - Lvl {lvl}";
                        
                        double area = slab["area"]?.Value<double>() ?? 0.0;
                        outAreas.Append(new GH_Number(area), path);
                        outLabels.Append(new GH_String(label), path);

                        double trueZ = slab["true_z"]?.Value<double>() ?? 0.0;
                        Transform zTranslation = Transform.Translation(0, 0, trueZ);

                        JArray boundaries = slab["boundary"] as JArray;
                        if (boundaries != null)
                        {
                            foreach (JToken crvData in boundaries)
                            {
                                Curve c = DeserializeCurve(crvData as JArray);
                                if (c != null)
                                {
                                    c.Transform(zTranslation);
                                    outBounds.Append(new GH_Curve(c), path);
                                }
                            }
                        }

                        matchCount++;
                        slabIndex++;
                    }
                    bldgIndex++;
                }

                sw.Stop();
                string searchMode = exactToggle ? "Exact" : "Flexible";
                this.Message = $"{this.NickName}\nTime: {sw.ElapsedMilliseconds} ms\n---\nReturned: {matchCount}\nMode: {searchMode}";
                
                DA.SetDataTree(0, outBounds);
                DA.SetDataTree(1, outAreas);
                DA.SetDataTree(2, outLabels);
            }
            catch (Exception ex)
            {
                this.Message = "JSON Parse Error:\n" + ex.Message;
                DA.SetDataTree(0, outBounds);
                DA.SetDataTree(1, outAreas);
                DA.SetDataTree(2, outLabels);
            }
        }

        private bool IsMatch(string targetName, List<string> filterList, bool exactMode)
        {
            if (filterList == null || filterList.Count == 0) return true;
            if (string.IsNullOrEmpty(targetName)) return false;

            string target = targetName.Trim().ToUpper();

            foreach (string f in filterList)
            {
                string pattern = f.Trim().ToUpper();
                if (exactMode)
                {
                    if (target == pattern) return true;
                }
                else
                {
                    if (pattern.Contains("*") || pattern.Contains("?"))
                    {
                        string regexPattern = "^" + Regex.Escape(pattern).Replace("\\*", ".*").Replace("\\?", ".") + "$";
                        if (Regex.IsMatch(target, regexPattern)) return true;
                    }
                    else if (target.Contains(pattern))
                    {
                        return true;
                    }
                }
            }
            return false;
        }

        private Curve DeserializeCurve(JArray segmentsData)
        {
            if (segmentsData == null || segmentsData.Count == 0) return null;

            List<Curve> crvs = new List<Curve>();

            foreach (JToken segToken in segmentsData)
            {
                JObject seg = segToken as JObject;
                if (seg == null) continue;

                string stype = seg["type"]?.Value<string>();
                if (stype == "Line")
                {
                    Point3d start = ParsePoint(seg["start"] as JArray);
                    Point3d end = ParsePoint(seg["end"] as JArray);
                    crvs.Add(new LineCurve(start, end));
                }
                else if (stype == "Arc")
                {
                    Point3d start = ParsePoint(seg["start"] as JArray);
                    Point3d mid = ParsePoint(seg["mid"] as JArray);
                    Point3d end = ParsePoint(seg["end"] as JArray);
                    crvs.Add(new ArcCurve(new Arc(start, mid, end)));
                }
                else if (stype == "Polyline")
                {
                    JArray pointsArray = seg["points"] as JArray;
                    if (pointsArray != null)
                    {
                        List<Point3d> pts = new List<Point3d>();
                        foreach (JToken ptToken in pointsArray)
                        {
                            pts.Add(ParsePoint(ptToken as JArray));
                        }
                        if (pts.Count > 0)
                        {
                            crvs.Add(new PolylineCurve(pts));
                        }
                    }
                }
            }

            if (crvs.Count == 0) return null;
            if (crvs.Count == 1) return crvs[0];

            Curve[] joined = Curve.JoinCurves(crvs, 0.01);
            if (joined != null && joined.Length > 0)
            {
                Curve crv = joined[0];
                if (!crv.IsClosed) crv.MakeClosed(0.01);
                return crv;
            }

            return null;
        }

        private Point3d ParsePoint(JArray arr)
        {
            if (arr == null || arr.Count < 3) return Point3d.Origin;
            return new Point3d(
                arr[0].Value<double>(),
                arr[1].Value<double>(),
                arr[2].Value<double>()
            );
        }

        protected override System.Drawing.Bitmap Icon
        {
            get
            {
                return IconLoader.Load("Slab_Parser.png");
            }
        }

        public override Guid ComponentGuid
        {
            get { return new Guid("11111111-2222-3333-4444-555555555556"); }
        }
    }
}

File: ScriptsToProcess/Slab_Parser.py
"""
SLAB JSON PARSER
================================================================================
INPUTS:
    Slab_JSON       (str)  [Item Access]
    Filter_Building (str)  [List Access]
    Filter_Tower    (str)  [List Access]
    Filter_Level    (int)  [List Access]
    ExactMatch      (bool) [Item Access]

OUTPUTS:
    SlabBounds, Areas, Labels
================================================================================
"""
import json
import Rhino.Geometry as rg
import Grasshopper as gh
from Grasshopper.Kernel.Data import GH_Path
import System
import fnmatch
import time

def update_ui(msg):
    _ghenv = globals().get('ghenv')
    _comp = globals().get('component')
    if _ghenv: _ghenv.Component.Message = msg
    elif _comp: _comp.Message = msg

_ghenv = globals().get('ghenv')
if _ghenv:
    _ghenv.Component.Name = "Slab JSON Parser"
    _ghenv.Component.NickName = "Slab_Parser"

def deserialize_curve(segments_data):
    if not segments_data: return None
    crvs = []
    for seg in segments_data:
        stype = seg.get("type")
        if stype == "Line": crvs.append(rg.LineCurve(rg.Point3d(*seg["start"]), rg.Point3d(*seg["end"])))
        elif stype == "Arc": crvs.append(rg.ArcCurve(rg.Arc(rg.Point3d(*seg["start"]), rg.Point3d(*seg["mid"]), rg.Point3d(*seg["end"]))))
        elif stype == "Polyline": crvs.append(rg.PolylineCurve([rg.Point3d(*p) for p in seg["points"]]))
    if not crvs: return None
    if len(crvs) == 1: return crvs[0]
    joined = rg.Curve.JoinCurves(crvs, 0.01)
    if joined and len(joined) > 0:
        crv = joined[0]
        if not crv.IsClosed: crv.MakeClosed(0.01)
        return crv
    return None

def is_match(target_name, filter_list, exact_mode):
    if not filter_list: return True 
    target = str(target_name).strip().upper()
    for f in filter_list:
        pattern = str(f).strip().upper()
        if exact_mode:
            if target == pattern: return True
        else:
            if '*' in pattern or '?' in pattern:
                if fnmatch.fnmatch(target, pattern): return True
            elif pattern in target:
                return True
    return False

def query_slabs():
    exec_start = time.perf_counter()
    json_in = globals().get('Slab_JSON', "")
    exact_toggle = globals().get('ExactMatch')
    exact_toggle = bool(exact_toggle) if exact_toggle is not None else False
    
    raw_bldg = globals().get('Filter_Building')
    f_bldg = [b for b in raw_bldg if b] if raw_bldg else []
    
    raw_tower = globals().get('Filter_Tower')
    f_tower = [t for t in raw_tower if t] if raw_tower else []
    
    raw_level = globals().get('Filter_Level')
    f_level = []
    if raw_level:
        for lvl in raw_level:
            if lvl is not None:
                try: f_level.append(int(lvl))
                except: pass
            
    out_bounds = gh.DataTree[System.Object]()
    out_areas = gh.DataTree[System.Object]()
    out_labels = gh.DataTree[System.Object]()
    
    if not json_in:
        update_ui("SLAB PARSER\nTime: 0.0 ms\n---\nAwaiting Data")
        return out_bounds, out_areas, out_labels
        
    try:
        data = json.loads(json_in)
        bldg_index = 0
        match_count = 0
        for bldg_name, slabs in data.items():
            if not is_match(bldg_name, f_bldg, exact_toggle): continue
            for slab_index, slab in enumerate(slabs):
                lvl = slab.get("floor_index", -1)
                tower_id = slab.get("tower_id", "Unknown")
                if f_level and lvl not in f_level: continue
                if not is_match(tower_id, f_tower, exact_toggle): continue
                    
                path = GH_Path(bldg_index, slab_index)
                label = f"{bldg_name} | {tower_id} - Lvl {lvl}"
                out_areas.Add(slab.get("area", 0.0), path)
                out_labels.Add(label, path)
                
                z_translation = rg.Transform.Translation(0, 0, slab.get("true_z", 0.0))
                for crv_data in slab.get("boundary", []):
                    c = deserialize_curve(crv_data)
                    if c: 
                        c.Transform(z_translation)
                        out_bounds.Add(c, path)
                match_count += 1
            bldg_index += 1
            
        exec_time = (time.perf_counter() - exec_start) * 1000
        search_mode = "Exact" if exact_toggle else "Flexible"
        update_ui(f"SLAB PARSER\nTime: {exec_time:.0f} ms\n---\nReturned: {match_count}\nMode: {search_mode}")
        return out_bounds, out_areas, out_labels
    except Exception as e:
        update_ui("JSON Parse Error:\n" + str(e))
        return out_bounds, out_areas, out_labels

bnds, areas, lbls = query_slabs()
SlabBounds = bnds
Areas = areas
Labels = lbls